keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,173 @@
+
+The search should be optimized (binary search) and preserve the proportion of pods lost across the variable count PodSets.
+
+The accepted number of pods in each PodSet are recorded in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`
+
+In order to evaluate the potential success of the preemption, the preemption process should be split in:
+- Target selection, which should select the workloads within cohort that will need to be evicted, if the list is empty the assignment will be treated as `NoFit`. This step will take place during nomination.
+- Preemption issuing, will do the actual eviction of the targets previously selected.
+
+**Partial admission with multiple variable PodSets count**
+
+If multiple PodSets within a workload have variable counts the way the counts should be decrease is highly dependent on the job framework needs, currently the following is proposed:
+
+Starting with the PodSets:
+
+```go
+[]podSet { {Count: 1} , {Count: 4, MinCount: 2}, {Count: 20, MinCount: 10}}
+```
+
+And only being able to admit 19 pods, it should end up with
+
+
+```go
+[]podSet { {Count: 1} , {Count: 3}, {Count: 15}}
+```
+
+With both pods that accept partial admission losing 50% of their range.
+
+The presented solution is not taking into account which of the variable count PodSets is causing the "NoFit" status and can potentially decrease the count of PodSets that will otherwise fit. If this behaviour is not suitable for a specific job framework, the integration layer of the framework can take into account limiting the variable count PodSets to 1.
+
+### Jobframework
+
+```diff
+type GenericJob interface {
+    // ...
+
+	
+-    // RunWithNodeAffinity will inject the node affinity extracting from workload to job and unsuspend the job.
++    // RunWithPodSetsInfo will inject the node affinity and podsSet counts extracted from workload to the job and unsuspend the job.
+-    RunWithNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RunWithPodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+-    // RestoreNodeAffinity will restore the original node affinity of job.
++    // RestorePodSetsInfo will restore the original node affinity of job.
+-    RestoreNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RestorePodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+
+    // ...
+}
+
+```
+
+### batch/Job controller
+
+Besides adapting `RunWithPodSetsInfo` and `RestorePodSetsInfo` it should also:
+
+- rework `PodSets()` to populate `MinCount` if the job is marked to support partial admission.
+  * jobs supporting partial admission should have a dedicated annotation. eg. `kueue.x-k8s.io/job-min-parallelism`, indicating the minimum `parallelism` acceptable by the job in case of partial admission.
```

Comment:
For native Job, I think we can simply return the minCount=1 at first, because native Job will run as far as possible.

Reply:
I don't think that would be acceptable to every customer. You might not be ok waiting months for your job to complete. You would rather wait in the queue for some time to have some level of parallelism available.

Reply:
Yes, defaulting `minCount=1` is rude, people may want to speed up the jobs with a big parallelism in a well-resourced cluster. The last concern is about the  growing annotations. 😀

Reply:
I agree, I'm also getting concerned about that. I hope we can push upstream some fields in the future.